packages/fuels-core/src/types/transaction_builders.rs

```diff
@@ -337,48 +328,64 @@ impl ScriptTransactionBuilder {
+            .take(num_witnesses as usize + unresolved_witnesses_len)
             .collect()
     }
 
-    // When the `script_gas_limit` was not set by the user, `dry_run` the tx
-    // and set the `script_gas_limit` to the actual `gas_used`
-    async fn set_script_gas_limit_to_gas_used(
+    fn no_input_to_pay_for_fee<'a, I: IntoIterator<Item = &'a FuelInput>>(inputs: I) -> bool {
+        !inputs.into_iter().any(|i| {
+            matches!(
+                i,
+                FuelInput::CoinSigned(_)
+                    | FuelInput::CoinPredicate(_)
+                    | FuelInput::MessageCoinSigned(_)
+                    | FuelInput::MessageCoinPredicate(_)
+            )
+        })
+    }
+
+    async fn gas_used(
         tx: &mut Script,
         provider: &impl DryRunner,
         network_info: &NetworkInfo,
         tolerance: f32,
-    ) -> Result<()> {
+    ) -> Result<u64> {
         let consensus_params = &network_info.consensus_parameters;
-        // Add `1` because of rounding
-        let max_gas = tx.max_gas(consensus_params.gas_costs(), consensus_params.fee_params()) + 1;
-
-        // TODO: @xgreenx why do I need to / 2
-        tx.set_script_gas_limit((network_info.max_gas_per_tx() / 2) - max_gas);
 
         // The `dry_run` validation will check if there is an input present that can cover
         // the tx fees. If we are estimating without inputs we have to add a temporary one
-        tx.inputs_mut().push(FuelInput::coin_signed(
-            Default::default(),
-            Default::default(),
-            1_000_000_000,
-            Default::default(),
-            TxPointer::default(),
-            0,
-            0u32.into(),
-        ));
+        let no_input_for_fee = Self::no_input_to_pay_for_fee(tx.inputs());
+        if no_input_for_fee {
+            tx.inputs_mut().push(FuelInput::coin_signed(
+                Default::default(),
+                Default::default(),
+                1_000_000_000,
+                Default::default(),
+                TxPointer::default(),
+                0,
+                0u32.into(),
+            ));
+
+            // Add an empty `Witness` for the `coin_signed` we just added
+            // and increase the witness limit
+            tx.witnesses_mut().push(Default::default());
+            tx.set_witness_limit(tx.witness_limit() + WITNESS_STATIC_SIZE as u64);
+        }
+
+        // Add `1` because of rounding
+        let max_gas = tx.max_gas(consensus_params.gas_costs(), consensus_params.fee_params()) + 1;
+        tx.set_script_gas_limit(network_info.max_gas_per_tx() - max_gas);
 
         let gas_used = provider
             .dry_run_and_get_used_gas(tx.clone().into(), tolerance)
             .await?;
 
-        // Remove the temporary coin
-        tx.inputs_mut().pop();
-
-        tx.set_script_gas_limit(gas_used);
+        // Remove dry_run input
+        if no_input_for_fee {
+            tx.inputs_mut().pop();
```

Comment:
Shouldn't we also reset the witnesses, witness limit and script_gas_limit here?

Reply:
We do not have to as we used the temporary `dry_run` witnesses. I will add a comment

Reply:
I just realized that we should readjust the witness_limit! So thanks for the question!

Reply:
Refactored the code again. Please have a look